### PR TITLE
Fix `cos_sin` device issue in Falcon model

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

We should update the device accordingly. See the comments along the changes.

So far, running the model on GPU, then change model to CPU and running CPU inputs, the cached tensors (like `cos_cached`) will be on GPU and fail to run with CPU tensors.

### CODE REVIEW

File: src/transformers/models/falcon/modeling_falcon.py [hunk 1]
```diff
@@ -129,6 +129,12 @@ def cos_sin(
         total_length = seq_len + past_key_values_length
         if total_length > self.seq_len_cached:
             self._set_cos_sin_cache(total_length, device, dtype)
+
+        # the cached tensors need to update their devices (for example, after we change the model's device)
+        if device != self.cos_cached:
+            self.cos_cached = self.cos_cached.to(device)
+            self.sin_cached = self.sin_cached.to(device)
```
Comment (ydshieh) [inline src/transformers/models/falcon/modeling_falcon.py:136]: without this, we have problems when switching a model device

Comment (ydshieh) [reply src/transformers/models/falcon/modeling_falcon.py:136]: Yes, don't know why I made such mistake (as it works on the code snippet I am using 😅  sorry). You saved me!

File: src/transformers/models/falcon/modeling_falcon.py [hunk 2]
```diff
@@ -129,6 +129,11 @@ def cos_sin(
         total_length = seq_len + past_key_values_length
         if total_length > self.seq_len_cached:
             self._set_cos_sin_cache(total_length, device, dtype)
+
+        # the cached tensors need to update their devices (for example, after we change the model's device)
+        self.cos_cached = self.cos_cached.to(device)
+        self.sin_cached = self.sin_cached.to(device)
```
Comment (ydshieh) [inline src/transformers/models/falcon/modeling_falcon.py:135]: I removed the check of device. As we can't compare device (given by a tensor) with a string directly.
If it is on the same device (inferred by torch itself), there won't be any operation performed.

**Review by LysandreJik — APPROVED** [review]
Great, thanks @ydshieh!

**Pull request merged**